File: Java-Emerson/rbTree/RbTree.java
package rbTree;

/*
    Aluno: Emerson Henrique Sulpino de Araújo
    Email: emerson.henrique.sulpino.araujo@ccc.ufcg.edu.br
    Matrícula: 123210141
    User do GitHub: Emerson349
*/

public class RbTree {
    private Node root; // Nó raiz da árvore
    private Node nil; // Nó NIL (representa folhas)

    public RbTree() {
        nil = new Node(0);
        nil.color = false; // NIL sempre preto
        root = nil; // Inicialmente, a árvore está vazia
    }

    public void insert(int value) {
        Node newNode = new Node(value);
        newNode.left = nil; // Novos nós apontam para NIL
        newNode.right = nil;
        newNode.father = null;

        root = insertRec(newNode, root); // Insere recursivamente
        balanceTree(newNode); // Balanceia a árvore após inserção
    }

    private Node insertRec(Node newNode, Node current) {
        if (current == nil) return newNode; // Caso base: inserção na folha

        if (newNode.value < current.value) {
            current.left = insertRec(newNode, current.left);
            current.left.father = current;
        } else if (newNode.value > current.value) {
            current.right = insertRec(newNode, current.right);
            current.right.father = current;
        }
        return current;
    }

    // Requilibrar arvore após inserção de um nó
    private void balanceTree(Node newNode) {
        while (newNode.father != null && newNode.father.color) {
            Node father = newNode.father;
            Node parent = father.father;

            if (father == parent.right) { // Caso o pai seja filho direito
                Node uncle = parent.left;

                if (uncle.color) { // Caso 1: tio vermelho
                    uncle.color = false;
                    father.color = false;
                    parent.color = true;
                    newNode = parent;
                } else {
                    if (newNode == parent.left) { // Caso 2: rotação à direita
                        newNode = father;
                        rotateRight(father);
                    }
                    father.color = false; // Caso 3: rotação à esquerda
                    parent.color = true;
                    rotateLeft(parent);
                }
            } else { // Caso simétrico (pai é filho esquerdo)
                Node uncle = parent.right;
                if (uncle.color) {
                    uncle.color = false;
                    father.color = false;
                    parent.color = true;
                    newNode = parent;
                } else {
                    if (newNode == parent.right) {
                        newNode = father;
                        rotateLeft(father);
                    }
                    father.color = false;
                    parent.color = true;
                    rotateRight(parent);
                }
            }
        }
        root.color = false; // Raiz sempre preta
    }

    private Node searchNode(int value, Node node) {
        if (node == nil) return nil; // Caso base: não encontrado
        if (value == node.value) return node;
        return value < node.value ? searchNode(value, node.left) : searchNode(value, node.right);
    }

    public void remove(int value) {
        Node nodeToRemove = searchNode(value, root);
        if (nodeToRemove == nil) return; // Se não encontrado, não faz nada
        deleteNode(nodeToRemove);
    }

    private void deleteNode(Node node) {
        Node x, y = node;
        boolean yOriginalColor = y.color;

        if (node.left == nil) { // Caso 1: sem filho esquerdo
            x = node.right;
            transplant(node, node.right);
        } else if (node.right == nil) { // Caso 2: sem filho direito
            x = node.left;
            transplant(node, node.left);
        } else { // Caso 3: nó com dois filhos
            y = successor(node.right);
            yOriginalColor = y.color;
            x = y.right;
            if (y.father == node) {
                x.father = y;
            } else {
                transplant(y, y.right);
                y.right = node.right;
                y.right.father = y;
            }
            transplant(node, y);
            y.left = node.left;
            y.left.father = y;
            y.color = node.color;
        }
        if (!yOriginalColor) {
            balanceDelete(x); // Chama o método de balanceamento da remoção
        }
    }

    private Node successor(Node node) {
        while (node.left != nil) { // Encontra o menor nó na subárvore direita
            node = node.left;
        }
        return node;
    }

    // substitui uma subárvore enraizada em um nó u por uma subárvore enraizada em um nó v
    private void transplant(Node u, Node v) {
        if (u.father == null)
            root = v;
        else if (u == u.father.left)
            u.father.left = v;
        else
            u.father.right = v;
        if (v != null) v.father = u.father;
    }

    // Reequilibrar a arvore após a remoção de um nó
    private void balanceDelete(Node x) {
        while (x != root && !x.color) {
            if (x == x.father.left) { // Caso 1: x é filho esquerdo
                Node w = x.father.right; // irmão vai ser o filho direito
                if (w.color) { // Caso 1.1 o irmão é vermelho
                    w.color = false;
                    x.father.color = true;
                    rotateLeft(x.father);
                    w = x.father.right;
                }
                if (!w.left.color && !w.right.color) { // Caso 1.2 ambos filhos do irmão são pretos
                    w.color = true;
                    x = x.father;
                } else { // Caso 1.3 Pelo menos um dos filhos do irmão não é preto
                    if (!w.right.color) {
                        w.left.color = false;
                        w.color = true;
                        rotateRight(w);
                        w = x.father.right;
                    }
                    w.color = x.father.color;
                    x.father.color = false;
                    w.right.color = false;
                    rotateLeft(x.father);
                    x = root;
                }
            } else { // Caso 2: x é filho direito
                Node w = x.father.left; // então o irmão será filho esquerdo
                // O caso 2 vai ter operações simétricas ao caso 1, mas, com rotações e verificações invertidas
                if (w.color) {
                    w.color = false;
                    x.father.color = true;
                    rotateRight(x.father);
                    w = x.father.left;
                }
                if (!w.right.color && !w.left.color) {
                    w.color = true;
                    x = x.father;
                } else {
                    if (!w.left.color) {
                        w.right.color = false;
                        w.color = true;
                        rotateLeft(w);
                        w = x.father.left;
                    }
                    w.color = x.father.color;
                    x.father.color = false;
                    w.left.color = false;
                    rotateRight(x.father);
                    x = root;
                }
            }
        }
        x.color = false;
    }

    // Rotação à esquerda
    private void rotateLeft(Node node) {
        Node right = node.right;
        node.right = right.left;
        if (right.left != nil) {
            right.left.father = node; // Atualiza o pai do filho esquerdo do filho direito
        }
        right.father = node.father;
        if (node.father == null) {
            root = right; // Se 'node' era raiz, o filho direito vira raiz
        } else if (node == node.father.left) {
            node.father.left = right; // Filho direito vira filho esquerdo do pai de 'node'
        } else {
            node.father.right = right; // Filho direito vira filho direito do pai de 'node'
        }
        right.left = node;
        node.father = right;
    }

    // Rotação à direita
    private void rotateRight(Node node) {
        Node left = node.left;
        node.left = left.right;
        if (left.right != nil) { // Atualiza o pai do filho direito do filho esquerdo
            left.right.father = node;
        }
        left.father = node.father;
        if (node.father == null) {
            root = left; // Se 'node' era raiz, o filho esquerdo vira raiz
        } else if (node == node.father.right) {
            node.father.right = left; // Filho esquerdo vira filho direito do pai de 'node'
        } else {
            node.father.left = left; // Filho esquerdo vira filho esquerdo do pai de 'node'
        }
        left.right = node;
        node.father = left;
    }

    // Método que usei para testes, mas é uma boa representação visual da àrvore
    public void printTree() {
        printTree(root, 0);
    }

    private void printTree(Node node, int level) {
        if (node == nil) {
            return;
        }
        printTree(node.right, level + 1);
        for (int i = 0; i < level; i++) {
            System.out.print("    ");
        }
        System.out.println(node.value + " (" + (node.color ? "RED" : "BLACK") + ")");
        printTree(node.left, level + 1);
    }

    // Classe Node
    private class Node {
        int value;
        Node left, right, father;
        boolean color; // true para vermelho, false para preto

        public Node(int value) {
            this.value = value;
            this.color = true; // Novos nós são sempre vermelhos
        }
    }
}

File: C++-Denis/RBTree.cc
#include <iostream>
#include <iomanip>

using namespace std;

/*
    Aluno: Denis Almeida Ferreira
    Email: denis.almeida.ferreira@ccc.ufcg.edu.br
    Matrícula: 123210975
    Usuário no GitHub: DenisAlmeidaFerreira

    Implementação de uma Árvore Rubro Negro
*/

enum Color { RED, BLACK };

class RBTree {
private:
    struct Node {
        int value;
        Node *left, *right, *father;
        Color color;

        Node(int val) : value(val), left(nullptr), right(nullptr), father(nullptr), color(RED) {}
    };

    Node *root, *nil;

    void rotateLeft(Node *node) {
        Node *right = node->right;
        node->right = right->left;
        if (right->left != nil) right->left->father = node;
        right->father = node->father;
        if (node->father == nullptr) root = right;
        else if (node == node->father->left) node->father->left = right;
        else node->father->right = right;
        right->left = node;
        node->father = right;
    }

    void rotateRight(Node *node) {
        Node *left = node->left;
        node->left = left->right;
        if (left->right != nil) left->right->father = node;
        left->father = node->father;
        if (node->father == nullptr) root = left;
        else if (node == node->father->right) node->father->right = left;
        else node->father->left = left;
        left->right = node;
        node->father = left;
    }

    void balanceTree(Node *newNode) {
        while (newNode->father && newNode->father->color == RED) {
            Node *father = newNode->father;
            Node *parent = father->father;
            if (father == parent->right) {
                Node *uncle = parent->left;
                if (uncle->color == RED) {
                    uncle->color = BLACK;
                    father->color = BLACK;
                    parent->color = RED;
                    newNode = parent;
                } else {
                    if (newNode == father->left) {
                        newNode = father;
                        rotateRight(father);
                    }
                    father->color = BLACK;
                    parent->color = RED;
                    rotateLeft(parent);
                }
            } else {
                Node *uncle = parent->right;
                if (uncle->color == RED) {
                    uncle->color = BLACK;
                    father->color = BLACK;
                    parent->color = RED;
                    newNode = parent;
                } else {
                    if (newNode == father->right) {
                        newNode = father;
                        rotateLeft(father);
                    }
                    father->color = BLACK;
                    parent->color = RED;
                    rotateRight(parent);
                }
            }
        }
        root->color = BLACK;
    }

    Node* insertRec(Node *newNode, Node *current) {
        if (current == nil) return newNode;
        if (newNode->value < current->value) {
            current->left = insertRec(newNode, current->left);
            current->left->father = current;
        } else if (newNode->value > current->value) {
            current->right = insertRec(newNode, current->right);
            current->right->father = current;
        }
        return current;
    }

    void printTree(Node* root, int indent = 0) {
        if (root == nullptr || root == nil) return;

        printTree(root->right, indent + 4);

        cout << setw(indent) << " " << root->value << " (" << (root->color == RED ? "RED" : "BLACK") << ")\n";

        printTree(root->left, indent + 4);
    }

    void transplant(Node *target, Node *replacement) {
        if (target->father == nullptr) root = replacement;
        else if (target == target->father->left) target->father->left = replacement;
        else target->father->right = replacement;
        replacement->father = target->father;
    }

    Node* findMinimum(Node *node) {
        while (node->left != nil) node = node->left;
        return node;
    }

    void fixDeletion(Node *node) {
        while (node != root && node->color == BLACK) {
            if (node == node->father->left) {
                Node *sibling = node->father->right;
                if (sibling->color == RED) {
                    sibling->color = BLACK;
                    node->father->color = RED;
                    rotateLeft(node->father);
                    sibling = node->father->right;
                }
                if (sibling->left->color == BLACK && sibling->right->color == BLACK) {
                    sibling->color = RED;
                    node = node->father;
                } else {
                    if (sibling->right->color == BLACK) {
                        sibling->left->color = BLACK;
                        sibling->color = RED;
                        rotateRight(sibling);
                        sibling = node->father->right;
                    }
                    sibling->color = node->father->color;
                    node->father->color = BLACK;
                    sibling->right->color = BLACK;
                    rotateLeft(node->father);
                    node = root;
                }
            } else {
                Node *sibling = node->father->left;
                if (sibling->color == RED) {
                    sibling->color = BLACK;
                    node->father->color = RED;
                    rotateRight(node->father);
                    sibling = node->father->left;
                }
                if (sibling->right->color == BLACK && sibling->left->color == BLACK) {
                    sibling->color = RED;
                    node = node->father;
                } else {
                    if (sibling->left->color == BLACK) {
                        sibling->right->color = BLACK;
                        sibling->color = RED;
                        rotateLeft(sibling);
                        sibling = node->father->left;
                    }
                    sibling->color = node->father->color;
                    node->father->color = BLACK;
                    sibling->left->color = BLACK;
                    rotateRight(node->father);
                    node = root;
                }
            }
        }
        node->color = BLACK;
    }

public:
    RBTree() {
        nil = new Node(0);
        nil->color = BLACK;
        root = nil;
    }

    void insert(int value) {
        Node *newNode = new Node(value);
        newNode->left = nil;
        newNode->right = nil;
        root = insertRec(newNode, root);
        balanceTree(newNode);
    }

    void printTree() {
        printTree(root, 0);
    }

    void remove(int value) {
        Node *current = root, *successor, *fixNode;
        while (current != nil && current->value != value) {
            if (value < current->value) current = current->left;
            else current = current->right;
        }
        if (current == nil) return;

        successor = current;
        Color originalColor = successor->color;
        if (current->left == nil) {
            fixNode = current->right;
            transplant(current, current->right);
        } else if (current->right == nil) {
            fixNode = current->left;
            transplant(current, current->left);
        } else {
            successor = findMinimum(current->right);
            originalColor = successor->color;
            fixNode = successor->right;
            if (successor->father == current) fixNode->father = successor;
            else {
                transplant(successor, successor->right);
                successor->right = current->right;
                successor->right->father = successor;
            }
            transplant(current, successor);
            successor->left = current->left;
            successor->left->father = successor;
            successor->color = current->color;
        }
        delete current;
        if (originalColor == BLACK) fixDeletion(fixNode);
    }
};

int main() {
    RBTree rbTree;

    cout << "Inserindo valores..." << endl;
    rbTree.insert(10);
    rbTree.insert(20);
    rbTree.insert(30);
    rbTree.insert(40);
    rbTree.insert(50);
    rbTree.insert(25);

    cout << "\nÁrvore após inserções:" << endl;
    rbTree.printTree();

    cout << "\nRemovendo valores..." << endl;
    rbTree.remove(30);
    rbTree.remove(10);

    cout << "\nÁrvore após remoções:" << endl;
    rbTree.printTree();

    return 0;
}
